url7a45f192ca_mfullen_ping_pong_net_tgz-pJ8-ping_pong_net_client_io_TestClientNonPPNServerJ8/src/ping/pong/net/connection/io/IoTcpWriteRunnable.java

```diff
@@ -0,0 +1,135 @@
+    /**
+     * Logger for IoTcpReadRunnable
+     */
+    public static final Logger LOGGER = LoggerFactory.getLogger(IoTcpWriteRunnable.class);
+    /**
+     * The socket the thread is reading from
+     */
+    protected Socket tcpSocket = null;
+    /**
+     * The inputstream in which the thread is reading from
+     */
+    protected OutputStream outputStream = null;
+    /**
+     * Flag for whether this thread is running
+     */
+    protected boolean running = false;
+    /**
+     * The queue of messages to write from
+     */
+    protected BlockingQueue<MessageType> writeQueue = new LinkedBlockingQueue<MessageType>();
+    /**
+     * Notifies the listener when this runnable is closed
+     */
+    protected RunnableEventListener runnableEventListener = null;
+    private DataWriter dataWriter = null;
+    protected DisconnectState disconnectState = DisconnectState.NORMAL;
+
+    public IoTcpWriteRunnable(RunnableEventListener runnableEventListener, DataWriter dataWriter, Socket tcpSocket)
+    {
+        this.tcpSocket = tcpSocket;
+        this.runnableEventListener = runnableEventListener;
+        this.dataWriter = dataWriter;
+    }
+
+    /**
+     * Initialize the Output Stream from the socket
+     */
+    protected void init()
+    {
+        this.outputStream = dataWriter.init(this.tcpSocket);
+    }
+
+    /**
+     * Is this thread still running?
+     * @return
+     */
+    public boolean isRunning()
+    {
+        return this.running;
+    }
+
+    /**
+     * Closes the thread but properly shuts down the socket by closing the outputstream.
+     * This method flushes the output stream before closing
+     * and the tcpsocket. Calls Socket.close
+     */
+    public void close()
+    {
+        try
+        {
+            this.running = false;
+            this.outputStream.flush();
+            this.outputStream.close();
+            synchronized (tcpSocket)
+            {
+                this.tcpSocket.close();
+            }
+        }
+        catch (IOException ex)
+        {
+            LOGGER.error("Error closing Socket", ex);
+        }
+        finally
+        {
+            if (this.runnableEventListener != null)
+            {
+                this.runnableEventListener.onRunnableClosed(disconnectState);
+                this.runnableEventListener = null;
+            }
+        }
+    }
+
+    /**
+     * Enqueue a message to the write Queue
+     * @param message the message to queue
+     * @return true if successful, false if not
+     */
+    public boolean enqueueMessage(MessageType message)
+    {
+        return this.writeQueue.add(message);
+    }
+
+    @Override
+    public void run()
+    {
+        this.init();
+
+        this.running = true;
+
+        while (this.running)
+        {
+            try
+            {
+                MessageType message = this.writeQueue.take();
+                this.dataWriter.writeData(message);
+            }
+            catch (InterruptedException ex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)